### ISSUE TITLE
Content Structure

### ISSUE THREAD

**Issue by mikearnaldi** [Effect-TS/website#59]
This issue to begin discussion of the content structure of the website, I think we should pause accepting PRs and remove current content until the structure is agreed and we have a baseline.

The reason is fairly simple, while all the content may look good in small it hardly make any sense as a whole, for example:
- https://www.effect.website/docs/core-concepts/option
- https://www.effect.website/docs/core-concepts/either
- https://www.effect.website/docs/advanced-concepts/types

While each page may look nice if one is reading precisely about that topic none of it is any relevant to learning Effect, reading through the details of the page `Option` & `Either` there is twice the explainer of the `map` function, conceptually it shouldn't be there in the first place.

Starting with the purpose of the website doc, the website doc is meant to approach beginners and experts alike, it should be a beginning guide into the world of effect and not a replacement of reference docs, if I want to know all the typeclasses that are available in effect the place is `https://effect-ts.github.io/data/modules`, if I am looking at the website I want to understand the concepts and the "what's best".

I am by no mean a content writer but I'll start with a proposed structure:

```
- Core Concepts
  - Effect
  - Errors 
  - Services
  - Layers
  - Runtime
  - Batching & Caching
  - Logging
  - Metrics
  - Configurations
  - Testing
  - Schemas
  - Immutability
  - Equality
  
- Utility Modules
  - Option
  - Either
  - HashMap
  - HashSet
  - Chunk
  - ...
  
- Deep Dive
  - Cache
  - Cause
  - Deferred
  - Exit
  - Fiber
  - FiberRef
  - Pool
  - Scope
  - Queue
  - Hub
  - Random
  - Clock
  - ...
 ```
 
 Each module should introduce the concepts as new and should focus on articulating:
 
 - why should I use this module?
 - when should I use this module?
 - how should I use this module?
 - where can I look for further references?
 
 Generally I'd like to avoid copying from any pre-existing doc attempt (let it be in fp-ts repo or in my blog posts) and rather focus on producing something new thinking that we need to explain those concepts to a person who potentially doesn't even know TS well or else programming in any serious manner

**Comment by aniravi24** [comment]
I agree with this sentiment. I don't really think the stuff that was ported from fp-ts has a good place in the docs yet, and shouldn't have been included without clarity on where it should be.

For the core concepts, at minimum, I initially planned to have at least these topics.

1) creating effects
2) running effects
3) context/services
4) layers
5) scope
6) runtime

I think usages of scope (e.g. `acquireRelease`, etc.) should be included before runtime.

I think some of the core concepts you mentioned could be in a separate section - not sure what to call it, but it would live somewhere between core concepts and deep dive.

I like the questions that each page should seek to answer.

I'm not totally sure what kind of content you're envisioning under some of those topics (e.g. `immutability` and `equality`), so once we find a structure we can agree on, we can dive more into what to cover.

I agree that we should avoid copying from pre-existing doc attempts and try to produce something new. We can look for more inspiration from other sources later, rather than letting those attempts bias our initial ideas.

**Comment by jessekelly881** [comment]
This structure looks like a really good start. I think the fp-ts docs are important to have though(although, probably buried) just to act as a reference. If Option, Either, et al are going to be mentioned in the docs(or used in code snippets) and used throughout the codebase it seems useful to have an explanation of the ideas _somewhere_ so that you can link to an explanation or so that users can search for an explanation on the docs site if they are having issues. It certainly doesn't have to be part of the main guides though(they aren't relevant there).

**Comment by mikearnaldi** [comment]
`I think the fp-ts docs are super important to have though, even if it's just there as a reference` => for "as a reference" there is the reference doc

**Comment by jessekelly881** [comment]
Makes sense! I think your probably right.

**Comment by mikearnaldi** [comment]
updated content structure with a new section "Utility Modules" for modules such as "Either / Option / HashMap / etc"

**Comment by aniravi24** [comment]
Eventually if we have other top-level sections in the docs such as Guides (Tutorials), Ecosystem, etc., it might make sense to move all of the sections you proposed above under a section called "Learn" or something like that, so there's less crowding of options to select from when one initially opens the menu.

**Comment by OlaoluwaM** [comment]
I think it might be worthwhile to find other projects whose documentations we can draw inspiration from. For instance, the Tailwind docs are pretty good. Another might be the new React docs. The idea here is to learn and adapt what already works, y'know 😄.

**Comment by mikearnaldi** [comment]
> Here's a rough draft for a possible structure. Thoughts?
> 
> ```lisp
> Get Started
> ├── Introduction
> │   └── Ecosystem
> ├── What is Effect
> └── Quick Start? (For those already familiar with fp-ts)
>     └── Guide mapping fp-ts to effect and link to other tutorials
> Core Concepts
> ├── .
> ├── ..
> ├── ...
> └── ....
> SubTopic-1
> ├── .
> ├── ..
> └── ...
> SubTopic-2
> ├── .
> ├── ..
> └── ...
> Tutorials/Examples/Recipes
> Contributing
> FAQ
> ```

I wouldn't put fp-ts prominently, the docs mostly has to address people who don't know anything about fp

**Comment by mikearnaldi** [comment]
> I think it might be worthwhile to find other projects whose documentations we can draw inspiration from. For instance, the Tailwind docs are pretty good. Another might be the new React docs. The idea here is to learn and adapt what already works, y'know 😄.

That's for sure something but the risk is drawing inspiration from something with a different surface, for example React / tailwind / etc are only talking about a single thing, in Effect we have modules that do completely different things and ecosystem packages so not sure where to look for that would be relevant

**Comment by OlaoluwaM** [comment]
> That's for sure something but the risk is drawing inspiration from something with a different surface, for example React / tailwind / etc are only talking about a single thing, in Effect we have modules that do completely different things and ecosystem packages so not sure where to look for that would be relevant

I think then it might be reasonable to have the Effect-TS docs be just about Effect and the things that relate to it directly, core concepts, examples, etc.... Anything else should have it own docs. What do you think? This also has the added benefit of keeping the docs smaller overall, which, IMO, will make Effect _seem_ like it's easy to get started with quickly

**Comment by err0r500** [comment]
Personally, I liked ZIO docs structure when I discovered it. 
Especially the fact they were oriented in a "problem solving" way : you want to achieve X -> you can do that with Y (based on examples which provide opinionated best practices). 
This is great to get started and when one's confident, reference doc _alla_ fp-ts is nice.

**Comment by SandroMaglione** [comment]
Hi everyone,

I would propose some suggestions for the docs:
- Why not add the [Introduction to Effect, Michael Arnaldi, WorkerConf 2022 Dornbirn](https://www.youtube.com/watch?v=zrNr3JVUc8I) video in the docs? I think it helps a lot to frame the problems that Effect aims to solve and the actual solution in code 
- I think the `Examples` section plays an important role. The core of this section can contain some common usecases and how Effect helps to solve them,  and then this can also be subdivided in "Official examples" and "Third party examples" (similar to [Riverpod](https://docs-v2.riverpod.dev/docs/introduction))
  - For example, we may add the Effect wrapper of `fetch` from [Micheal's video](https://youtu.be/zrNr3JVUc8I?t=1611) as first example
- It may be worth adding each individual API reference as a sub-menu ([Sub docs](https://nextra.site/docs/docs-theme/page-configuration#navbar-items)) in the nav bar

If this works I can open a PR to add the structure above, as well as some other examples of Effect code that I implemented.

**Comment by ethanniser** [comment]
Hi all, recently discovered Effect and have really fell in love. I know **how important having good docs is for driving adoption** and so I want to help contribute because there is a lot of work to do.

When looking for some kind of "plan" or "structure" for the docs I found this issue, which is kind of dead and never really went too far in the first place.

I know @gcanti has been working hard on the docs and it is awesome work, I don't know if he has his own 'structure' he is working off of, but I think it would be helpful for everyone to have a clear plan for what we want the docs to look like. This will make it easy for contributors to browse what is left to be completed, request and be assigned it.

I spent the last hour making a fresh "content structure". 

**To start I just want to emphasize I am still new to Effect and totally open to changes, discussion and feedback, I just want to get the ball rolling.**

Here were my goals when writing it:

1. Make the experience for new devs the absolute best it can be, a clear logical page -> page -> page that gets a new dev familiar with the concept of effect and the basics
2. Make it easy for intermediate Effect devs to find the more advanced tools to solve their problems
3. Merge the original structure from @mikearnaldi at the start of this issue (which I like the overall idea of) and the current state of the site (has some well written content on it)
4. Make the 'plan' as wide as possible and after completing it we can just focus on filling in the content we are able to, section by section

[here it is](https://gist.github.com/ethanniser/c0c5090deb0e513da568272e9c965e6b)

---

some final thoughts: 

- trying to organize a library this big is really challenging, there is so much to cover and many different ways to introduce different topics/concepts
- I think the way [pigoz/effect-crashcourse](https://github.com/pigoz/effect-crashcourse) does the first 4 chapters is really really good and we should take inspiration
- I found [this tweet](https://twitter.com/jbmusso/status/1668881198867775488?s=20) very insightful, and I think it provides an interesting prospective for writing documentation. I think we should aim to get a new dev familiar with these “30-40” most common functions as quickly and as easily as we can.
- we should avoid making “dropdown” buttons ALSO pages, as its very confusing, especially when currently some (scheduling) are and others (the rest) are not
- the home/index page should have a massive **"Get Started"** button that takes you to the first page of the docs, right now its kind of confusing where to go and the button is very small
---

Writing docs is really hard (and really important), so I'm excited to contribute to help make Effect's docs the best we can

let me know what you think, thanks!

**Comment by adrian-gierakowski** [comment]
@ethanniser I believe the structure has been agreed but it was never mentioned here. See https://discord.com/channels/795981131316985866/1116789625017610312/1116822642499002470. The work has been progressing quite well with new content dropping daily.

**Comment by ethanniser** [comment]
I see

Really wish somehow had updated or closed this issue, I think its unreasonable to expected people new to Effect to be aware of information in non-pinned discord messages from weeks ago

All good though, still happy to help if help is wanted

**Comment by mikearnaldi** [comment]
> I see
> 
> Really wish someone had updated or closed this issue, I think its unreasonable to expected people new to Effect to be aware of information in non-pinned discord messages from weeks ago. Maybe we can put the roadmap somewhere in the repo?
> 
> All good though, still happy to help if help is wanted

That's my fault sorry about that, in order to unlock the first good draft of docs we've elected Giulio as the driver for this and it should have been mentioned in the issue. 

Contribution is welcome and if you're willing to help the effort please coordinate with Giulio on where to start, having a beginner mindset when writing docs is very important so I see the fact that you're just getting into Effect as a plus.

**Comment by gcanti** [comment]
> I found this tweet very insightful, and I think it provides an interesting prospective for writing documentation. I think we should aim to get a new dev familiar with these “30-40” most common functions as quickly and as easily as we can.

Yup, ideally these “30-40” most common functions should be covered by Effect Essentials + Tutorials

> we should avoid making “dropdown” buttons ALSO pages

I agree, the only exception **was** (https://github.com/Effect-TS/website/pull/93) scheduling

> the home/index page should have a massive button

:+1: https://github.com/Effect-TS/website/pull/94

(Note: we will work on a "real" landing page asap)

@ethanniser thanks for your suggestions

**Comment by gcanti** [comment]
@ethanniser great feedback, thank you. I agree with many points, but I would keep this issue focused on guidelines, the overall structure of the website, and (maybe) general matters that apply to all sections. I suggest moving the discussion on specific points to dedicated issues, one for each major topic (e.g., one for "Effect Essentials," one for "Error Management," etc.), so that they are neither too broad nor too granular with an excessive number of issues.

**Comment by ethanniser** [comment]
will make dedicated issues for the other stuff

------

as for this issue:

> Having well defined "categories" is great, although I still think it would be beneficial to have an actually written out plan for what actual pages we want to have in each of the sections. Maybe this already exists, but I can't find it.

still think this is important to do, if you agree

**Comment by ethanniser** [comment]
just want to bump this

I think planning out all the pages we eventually want to have and adding them as "stubs" would make the docs feel much more "complete" for no real work, as well as increasing the chance someone sees a stub for something they want to learn about and decides to help contribute to that section

**Comment by gcanti** [comment]
Yes, that's how we intend to proceed. We have envisioned a first series of pages and have added them as stubs (for example, [https://www.effect.website/docs/concepts/immutability](https://www.effect.website/docs/concepts/immutability)), and now we are filling them with content. Once this initial phase is completed, which also serves to refine the process and make many small decisions (such as what type of examples to include, how to indicate warnings, errors, etc.), we will have a new brainstorming session to start a new cycle of stubs + contents. The first cycle, which is currently underway, will take a bit longer because it's the first one and we need to test it concretely. However, if you already have some proposals for stubs for the second cycle, feel free to share them here.

**Comment by ethanniser** [comment]
Thank you for this response, this sounds like a good plan. Could we also add this to the readme @gcanti? I would consider this really important information for anyone wanting to contribute to the project that isn't mentioned anywhere in the repo. After looking back through the `#docs` channel in the discord today I found this message from @mikearnaldi 

> Giulio is working on a first version together with Johannes. The reason I paused uncoordinated effort is primarily because we were losing the direction too easily, with a focus group it's easier to maintain a coherent line. As soon as the first version is out we will have a great basis to contribute into

Totally understand this and I agree with the rationale, but as I mentioned earlier its just not reasonable to expect new contributors to be aware of information in non-pinned discord messages from weeks ago.

My actions as a contributor would've been very different having known this information, so I hope we can keep these important directional decisions reflected in the readme or a contributing guide going foward.

I look forward to contributing more once the 'first version' is complete

**Comment by gcanti** [comment]
> I hope we can keep these important directional decisions reflected in the readme or a contributing guide going foward

@ethanniser 👍  I have added a "Contributing" section to the README. Honestly, the work on the documentation has just started after a period of calm, and we were caught off guard. We didn't expect such quick participation (I mean, pleasantly surprised). Thank you for the feedback you have provided so far, it's highly valuable to us.

**Comment by IMax153** [comment]
Closing this discussion as the content structure has been relatively stable until the recent website rebuild.
